### PR TITLE
[wip] Remove CoffeeScript ref variables

### PR DESCRIPTION
**Pull request by kevinsawicki**
* Cleans up the `ref`, `ref1`, `ref2`, etc. variables left over from the CoffeeScript migration.
* Migrates to using `const` and `let` over long `var` lines at the top of several functions.
* Migrates to using `forEach`, `find`, `filter`, `map` over `for` loops in several places.

### Before

```js
BrowserWindow.fromWebContents = function(webContents) {
  var i, len, ref1, window, windows;
  windows = BrowserWindow.getAllWindows();
  for (i = 0, len = windows.length; i < len; i++) {
     window = windows[i];
     if ((ref1 = window.webContents) != null ? ref1.equal(webContents) : void 0) {
       return window;
     }
   }
};
```

### After

```js
BrowserWindow.fromWebContents = function(webContents) {
  return BrowserWindow.getAllWindows().find(function(window) {
    const contents = window.webContents;
    return contents != null && contents.equal(webContents);
  });
};
```

Refs #4065

### CODE REVIEW

**Comment by kevinsawicki** [comment]
This pull request is somewhat challenging to review due to this type of change so I'm going to review it several times myself and to make sure no unintended behavior was changed.

File: lib/browser/api/menu.js [hunk 1]
```diff
@@ -12,26 +12,29 @@ var nextGroupId = 0;
 // Search between separators to find a radio menu item and return its group id,
 // otherwise generate a group id.
 var generateGroupId = function(items, pos) {
-  var i, item, j, k, ref1, ref2, ref3;
   if (pos > 0) {
-    for (i = j = ref1 = pos - 1; ref1 <= 0 ? j <= 0 : j >= 0; i = ref1 <= 0 ? ++j : --j) {
```
Comment (kevinsawicki) [inline lib/browser/api/menu.js]: I decided to migrate this line given its lack of readability

File: lib/browser/api/menu.js [hunk 2]
```diff
@@ -12,26 +12,29 @@ var nextGroupId = 0;
...
       if (item.type === 'radio') {
         return item.groupId;
       }
       if (item.type === 'separator') {
         break;
       }
+      i--;
     }
   } else if (pos < items.length) {
-    for (i = k = ref2 = pos, ref3 = items.length - 1; ref2 <= ref3 ? k <= ref3 : k >= ref3; i = ref2 <= ref3 ? ++k : --k) {
```
Comment (kevinsawicki) [inline lib/browser/api/menu.js]: Also decided to migrate this line given its lack of readability

**Comment by YuriSolovyov** [comment]
Since this PR is against master, can't you use rest args for `arguments` handling?

File: lib/browser/api/app.js [hunk 3]
```diff
@@ -55,18 +55,14 @@ app.getAppPath = function() {
 };
 
 // Routes the events to webContents.
-var ref1 = ['login', 'certificate-error', 'select-client-certificate'];
-var fn = function(name) {
-  return app.on(name, function() {
-    var args, event, webContents;
-    event = arguments[0], webContents = arguments[1], args = 3 <= arguments.length ? slice.call(arguments, 2) : [];
+['login', 'certificate-error', 'select-client-certificate'].forEach(function(name) {
```
Comment (zcbenz) [inline lib/browser/api/app.js]: We can probably use the `for of` syntax for iterating array.

File: lib/browser/api/app.js [hunk 4]
```diff
@@ -55,18 +55,14 @@ app.getAppPath = function() {
...
+  app.on(name, function() {
```
Comment (zcbenz) [inline lib/browser/api/app.js]:
```js
app.on(name, function(event, webContents, ...args) {
```

**Comment by kevinsawicki** [comment]
I'm going to close this out for now and break this up into some smaller pull requests that focus on the following things individually:

- Using rest parameters
- Using destructured assignenment
- Removing temp `ref` variables

**Pull request closed without merging**